### PR TITLE
Add tooling for publishing OpenShift kubebuilder tools

### PR DESCRIPTION
**Pull request by JoelSpeed**
Envtest, a testing suite used in openshift/api and most controller-runtime based operators leverages archives from a GCS bucket to provide etcd and KAS binaries for local and integration testing.

As of 4.16, we are carrying patches in KAS that bring some upstream API features into OpenShift ahead of their general release in KAS, specifically the CEL net.IP and net.CIDR extensions.

To leverage these, both o/api and others will need binaries provided to be able to test CRDs leveraging these extensions.

This PR adds a go tool/script that extracts the binaries from a payload image, creates the required archives and then uploads them to an openshift GCS bucket.

We will need to run this periodically to create new releases when new binaries are needed, circa once a release.

We may also deprecate support for this when the API features reconverge (4.18), but it's worth keeping the tooling and setup around in case we need to leverage it again in the future

### CODE REVIEW

File: tools/Makefile [hunk 1]
```diff
@@ -27,6 +27,12 @@ clean:
 run-codegen: codegen
 	$(OUTPUT_DIR)/codegen --base-dir $(BASE_DIR) --api-group-versions $(API_GROUP_VERSIONS) --required-feature-sets $(OPENSHIFT_REQUIRED_FEATURESETS)
 
+PULL_SECRET?=${PULL_SECRET:-$(HOME)/pull-secret.txt}
+
+.PHONY: publish-kubebuilder-tools
+publish-kubebuilder-tools:
+	go run ./publish-kubebuilder-tools/main.go -version 1.29.1 -output-dir /tmp/kubebuilder-tools -payload=registry.ci.openshift.org/ocp/release:4.16.0-0.ci-2024-02-20-022620 -pull-secret $(PULL_SECRET)
```
Comment (deads2k) [inline tools/Makefile:34]: if memory serves, there's some way to have a repo build produce a change to an imagestream.  I'm ok with some ugliness to get our current problem resolved, but a thing to consider for a future update.

File: tests/Makefile [hunk 2]
```diff
@@ -8,7 +8,7 @@ integration: verify-vendor test
 
 .PHONY: test
 test: ## Run only the tests.
```
Comment (deads2k) [inline tests/Makefile:10]: Can we have the old version of this as a different target?  Say `test-legacy` or `test-upstream` or some such?  When this breaks, I'd like to get our repo mostly unstuck quickly.

File: tools/publish-kubebuilder-tools/main.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+	pullSecretFile := flag.String("pull-secret", "", "The pull secret to use for the kubebuilder tools")
+	version := flag.String("version", "", "The version of the kubebuilder tools to publish. This should be a Kubernetes version that the build is based upon.")
+	outputDir := flag.String("output-dir", "", "The output directory to write the kubebuilder tools to")
+	payload := flag.String("payload", "", "The payload to use for building the kubebuilder tools archives. This should be in the format registry.ci.openshift.org/ocp/release:<version>")
+	skipUpload := flag.Bool("skip-upload", false, "Skip uploading the artifacts created to the openshift-gce-devel/openshift-kubebuilder-tools bucket")
+
+	flag.Parse()
+
+	if *pullSecretFile == "" {
+		panic("pull-secret is required")
```
Comment (deads2k) [inline tools/publish-kubebuilder-tools/main.go:33]: Give me a hint for how to create this.  Do I log into a CI cluster and do the `oc registry login` thing?

File: tools/publish-kubebuilder-tools/main.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+)
+
+func main() {
+	pullSecretFile := flag.String("pull-secret", "", "The pull secret to use for the kubebuilder tools")
+	version := flag.String("version", "", "The version of the kubebuilder tools to publish. This should be a Kubernetes version that the build is based upon.")
```
Comment (deads2k) [inline tools/publish-kubebuilder-tools/main.go:25]: does it include z-streams or not?

File: tests/Makefile [hunk 5]
```diff
@@ -8,7 +8,7 @@ integration: verify-vendor test
...
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path --bin-dir $(PROJECT_DIR)/bin)" ./hack/test.sh
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path --bin-dir $(PROJECT_DIR)/bin --remote-bucket openshift-kubebuilder-tools)" ./hack/test.sh
```
Comment (deads2k) [inline tests/Makefile:11]: can I get a comment with a link to the docs for whatever command has this --remote-bucket arg so should you win the lottery I have some shot at reverse engineering how it works. :)

**Comment by deads2k** [comment]
I understand the goal.  I think it'll be fragile over time.  I'm ok doing this to resolve our initial problem, but in three releases or so we should try to collapse back.

Some comments.  I'd like the legacy-test target created for a rapid fallback.

/lgtm
/hold until we agree on IOU vs update

@JoelSpeed your credit in the area is good.  Your call on IOU or pre-merge.

File: tools/publish-kubebuilder-tools/main.go [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+	}
+
+	// Decrypt the pull secret to get the bearer token.
+	registryAuthToken, err := getRegistryAuthToken(*pullSecretFile)
+	if err != nil {
+		panic(err)
+	}
+
+	// We only expect images from the internal releases
+	if *payload == "" || !strings.HasPrefix(*payload, "registry.ci.openshift.org/ocp/release:") {
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply tools/publish-kubebuilder-tools/main.go:51]: I could update it to use any registry I guess, I just made some assumptions here that people would always use a nightly. Any timeline for that?

File: tools/publish-kubebuilder-tools/main.go (hunk 3 above)
Comment (JoelSpeed) [reply tools/publish-kubebuilder-tools/main.go:33]: Added a note to the makefile where this is being called

**Pull request merged**